qiskit_nature/problems/sampling/protein_folding/qubit_op_builder.py

```diff
@@ -0,0 +1,545 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""Builds qubit operators for all Hamiltonian terms in the protein folding problem."""
+from typing import List, Union
+
+from qiskit.opflow import OperatorBase, PauliOp, PauliSumOp
+
+from .bead_contacts.contact_map import ContactMap
+from .bead_distances.distance_map import DistanceMap
+from .exceptions.invalid_side_chain_exception import (
+    InvalidSideChainException,
+)
+from .exceptions.invalid_size_exception import InvalidSizeException
+from .penalty_parameters import PenaltyParameters
+from .peptide.pauli_ops_builder import _build_full_identity
+from .qubit_utils.qubit_fixing import _fix_qubits
+from .peptide.beads.base_bead import BaseBead
+from .peptide.peptide import Peptide
+
+
+# TODO link to a SamplingQubitOpBuilder interface
+def _build_qubit_op(
+    peptide: Peptide,
+    pair_energies: List[List[List[List[float]]]],
+    penalty_parameters: PenaltyParameters,
+    n_contacts: int,
+) -> Union[PauliSumOp, PauliOp]:
+    """
+    Builds a qubit operator for a total Hamiltonian for a protein folding problem. It includes
+    8 terms responsible for chirality, geometry and nearest neighbors interactions.
+
+    Args:
+        peptide: A Peptide object that includes all information about a protein.
+        pair_energies: Numpy array of pair energies for amino acids.
+        penalty_parameters: A PenaltyParameters object storing the values of all penalty
+                            parameters.
+        n_contacts: number of contacts between all beads.
+
+
+    Returns:
+        h_total: A total Hamiltonian for the protein folding problem.
+
+    Raises:
+        InvalidSizeException: if chains of invalid/incompatible sizes provided.
+        InvalidSideChainException: if side chains on forbidden indices provided.
+    """
+    lambda_chiral, lambda_back, lambda_1, lambda_contacts = (
+        penalty_parameters.lambda_chiral,
+        penalty_parameters.lambda_back,
+        penalty_parameters.lambda_1,
+        penalty_parameters.lambda_contacts,
+    )
+    side_chain = peptide.get_side_chain_hot_vector()
+    main_chain_len = len(peptide.get_main_chain)
+
+    if len(side_chain) != main_chain_len:
+        raise InvalidSizeException("side_chain_lens size not equal main_chain_len")
```

Comment:
Maybe look over these error messages - having the failed values formatted into the message can be helpful for understanding what might have gone wrongs eg in this case. Also maybe we should say side chain size or something an user might more readily comprehend rather than using internal variable names as part of message
```
raise InvalidSizeException(f"side chain size: {len(side_chain)} not equal main chain size {main_chain_len}")
```